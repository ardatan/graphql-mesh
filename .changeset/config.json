{
  "$schema": "https://unpkg.com/@changesets/config@2.1.0/schema.json",
  "commit": false,
  "linked": [],
  "access": "restricted",
  "baseBranch": "master",
  "updateInternalDependencies": "patch",
  "ignore": ["website"],
  "changelog": ["@changesets/changelog-github", { "repo": "urigo/graphql-mesh" }],
  "snapshot": {
    "useCalculatedVersion": true,
    "prereleaseTemplate": "{tag}-{datetime}-{commit}"
  }
}
