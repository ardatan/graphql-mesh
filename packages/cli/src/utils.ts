export const stripJSONComments = (data: string) => {
  const re = /\/\/(.*)/g;
  const re2 = /\/\*(.*)\*\//g;
  return data.replace(re, '').replace(re2, '');
};
