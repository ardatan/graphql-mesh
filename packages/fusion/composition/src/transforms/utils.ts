export class TransformValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TransformValidationError';
  }
}
