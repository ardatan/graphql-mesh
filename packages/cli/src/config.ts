import { ConfigProcessOptions, processConfig } from '@graphql-mesh/config';
import { jsonSchema, YamlConfig } from '@graphql-mesh/types';
import { defaultImportFn, loadYaml } from '@graphql-mesh/utils';
import Ajv from 'ajv';
import { cosmiconfig, defaultLoaders } from 'cosmiconfig';
import path from 'path';

export function validateConfig(config: any): asserts config is YamlConfig.Config {
  const ajv = new Ajv({
    strict: false,
  } as any);
  jsonSchema.$schema = undefined;
  const isValid = ajv.validate(jsonSchema, config);
  if (!isValid) {
    console.warn(`Configuration is not valid:\n${ajv.errorsText()}`);
  }
}

export async function findAndParseConfig(options?: ConfigProcessOptions) {
  const { configName = 'mesh', dir: configDir = '', ...restOptions } = options || {};
  const dir = path.isAbsolute(configDir) ? configDir : path.join(process.cwd(), configDir);
  const explorer = cosmiconfig(configName, {
    loaders: {
      '.json': customLoader('json', options?.importFn),
      '.yaml': customLoader('yaml', options?.importFn),
      '.yml': customLoader('yaml', options?.importFn),
      '.js': customLoader('js', options?.importFn),
      '.ts': customLoader('js', options?.importFn),
      noExt: customLoader('yaml', options?.importFn),
    },
  });
  const results = await explorer.search(dir);

  if (!results) {
    throw new Error(`No ${configName} config file found in "${dir}"!`);
  }

  const config = results.config;
  validateConfig(config);
  return processConfig(config, { dir, ...restOptions });
}

function customLoader(ext: 'json' | 'yaml' | 'js', importFn = defaultImportFn) {
  function loader(filepath: string, content: string) {
    if (process.env) {
      content = content.replace(/\$\{(.*?)\}/g, (_, variable) => {
        let varName = variable;
        let defaultValue = '';

        if (variable.includes(':')) {
          const spl = variable.split(':');
          varName = spl.shift();
          defaultValue = spl.join(':');
        }

        return process.env[varName] || defaultValue;
      });
    }

    if (ext === 'json') {
      return defaultLoaders['.json'](filepath, content);
    }

    if (ext === 'yaml') {
      return loadYaml(filepath, content);
    }

    if (ext === 'js') {
      return importFn(filepath);
    }
  }

  return loader;
}
