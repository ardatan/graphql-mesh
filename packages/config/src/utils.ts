import { parse } from 'graphql';
import { MeshHandlerLibrary, KeyValueCache, YamlConfig, MergerFn, ImportFn, MeshPubSub } from '@graphql-mesh/types';
import { resolve } from 'path';
import { IResolvers, printSchemaWithDirectives } from '@graphql-tools/utils';
import { paramCase } from 'param-case';
import { loadTypedefs } from '@graphql-tools/load';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import { get, set, kebabCase } from 'lodash';
import { stringInterpolator } from '@graphql-mesh/utils';
import { mergeResolvers } from '@graphql-tools/merge';
import { PubSub, withFilter } from 'graphql-subscriptions';
import { EventEmitter } from 'events';
import { CodeFileLoader } from '@graphql-tools/code-file-loader';
import StitchingMerger from '@graphql-mesh/merger-stitching';

export async function getPackage<T>(name: string, type: string, importFn: ImportFn): Promise<T> {
  const casedName = paramCase(name);
  const casedType = paramCase(type);
  const possibleNames = [
    `@graphql-mesh/${casedName}`,
    `@graphql-mesh/${casedName}-${casedType}`,
    `@graphql-mesh/${casedType}-${casedName}`,
    casedName,
    `${casedName}-${casedType}`,
    `${casedType}-${casedName}`,
    casedType,
  ];
  if (name.includes('-')) {
    possibleNames.push(name);
  }
  const possibleModules = possibleNames.concat(resolve(process.cwd(), name));

  for (const moduleName of possibleModules) {
    try {
      const exported = await importFn(moduleName);

      return (exported.default || exported.parser || exported) as T;
    } catch (err) {
      if (
        !err.message.includes(`Cannot find module '${moduleName}'`) &&
        !err.message.includes(`Could not locate module`)
      ) {
        throw new Error(`Unable to load ${type} matching ${name}: ${err.message}`);
      }
    }
  }

  throw new Error(`Unable to find ${type} matching ${name}`);
}

export async function getHandler(name: keyof YamlConfig.Handler, importFn: ImportFn): Promise<MeshHandlerLibrary> {
  const handlerFn = await getPackage<MeshHandlerLibrary>(name.toString(), 'handler', importFn);

  return handlerFn;
}

export async function resolveAdditionalTypeDefs(baseDir: string, additionalTypeDefs: string) {
  if (additionalTypeDefs) {
    const sources = await loadTypedefs(additionalTypeDefs, {
      cwd: baseDir,
      loaders: [new CodeFileLoader(), new GraphQLFileLoader()],
    });
    return sources.map(source => source.document || parse(source.rawSDL || printSchemaWithDirectives(source.schema)));
  }
  return undefined;
}

export async function resolveAdditionalResolvers(
  baseDir: string,
  additionalResolvers: (
    | string
    | YamlConfig.AdditionalStitchingResolverObject
    | YamlConfig.AdditionalSubscriptionObject
  )[],
  importFn: ImportFn,
  pubsub: MeshPubSub
): Promise<IResolvers> {
  const loadedResolvers = await Promise.all(
    (additionalResolvers || []).map(async additionalResolver => {
      if (typeof additionalResolver === 'string') {
        const filePath = additionalResolver;

        const exported = await importFn(resolve(baseDir, filePath));
        let resolvers = null;

        if (exported.default) {
          if (exported.default.resolvers) {
            resolvers = exported.default.resolvers;
          } else if (typeof exported.default === 'object') {
            resolvers = exported.default;
          }
        } else if (exported.resolvers) {
          resolvers = exported.resolvers;
        }

        if (!resolvers) {
          console.warn(`Unable to load resolvers from file: ${filePath}`);

          return {};
        }

        return resolvers;
      } else {
        if ('pubsubTopic' in additionalResolver) {
          return {
            [additionalResolver.type]: {
              [additionalResolver.field]: {
                subscribe: withFilter(
                  (root, args, context, info) => {
                    const resolverData = { root, args, context, info };
                    const topic = stringInterpolator.parse(additionalResolver.pubsubTopic, resolverData);
                    return pubsub.asyncIterator(topic);
                  },
                  (root, args, context, info) => {
                    return additionalResolver.filterBy ? eval(additionalResolver.filterBy) : true;
                  }
                ),
                resolve: (payload: any) => {
                  if (additionalResolver.returnData) {
                    return get(payload, additionalResolver.returnData);
                  }
                  return payload;
                },
              },
            },
          };
        } else {
          return {
            [additionalResolver.type]: {
              [additionalResolver.field]: {
                selectionSet: additionalResolver.requiredSelectionSet,
                resolve: async (root: any, args: any, context: any, info: any) => {
                  const resolverData = { root, args, context, info };
                  const methodArgs: any = {};
                  for (const argPath in additionalResolver.args) {
                    set(methodArgs, argPath, stringInterpolator.parse(additionalResolver.args[argPath], resolverData));
                  }
                  const result = await context[additionalResolver.targetSource].api[additionalResolver.targetMethod](
                    methodArgs,
                    {
                      selectedFields: additionalResolver.resultSelectedFields,
                      selectionSet: additionalResolver.resultSelectionSet,
                      depth: additionalResolver.resultDepth,
                    }
                  );
                  return additionalResolver.returnData ? get(result, additionalResolver.returnData) : result;
                },
              },
            },
          };
        }
      }
    })
  );

  return mergeResolvers(loadedResolvers);
}

export async function resolveCache(
  cacheConfig: YamlConfig.Config['cache'],
  importFn: ImportFn
): Promise<KeyValueCache | undefined> {
  if (cacheConfig) {
    const cacheName = Object.keys(cacheConfig)[0];
    const config = cacheConfig[cacheName];

    const moduleName = kebabCase(cacheName.toString());
    const pkg = await getPackage<any>(moduleName, 'cache', importFn);
    const Cache = pkg.default || pkg;

    return new Cache(config);
  }
  const InMemoryLRUCache = await import('@graphql-mesh/cache-inmemory-lru').then(m => m.default);
  const cache = new InMemoryLRUCache();
  return cache;
}

export async function resolvePubSub(
  pubsubYamlConfig: YamlConfig.Config['pubsub'],
  importFn: ImportFn
): Promise<MeshPubSub> {
  if (pubsubYamlConfig) {
    let pubsubName: string;
    let pubsubConfig: any;
    if (typeof pubsubYamlConfig === 'string') {
      pubsubName = pubsubYamlConfig;
    } else {
      pubsubName = pubsubYamlConfig.name;
      pubsubConfig = pubsubYamlConfig.config;
    }

    const moduleName = kebabCase(pubsubName.toString());
    const pkg = await getPackage<any>(moduleName, 'pubsub', importFn);
    const PubSub = pkg.default || pkg;

    return new PubSub(pubsubConfig);
  } else {
    const eventEmitter = new EventEmitter({ captureRejections: true });
    eventEmitter.setMaxListeners(Infinity);
    const pubsub = new PubSub({ eventEmitter }) as MeshPubSub;

    return pubsub;
  }
}

export async function resolveMerger(mergerConfig: YamlConfig.Config['merger'], importFn: ImportFn): Promise<MergerFn> {
  if (mergerConfig) {
    const pkg = await getPackage<any>(mergerConfig, 'merger', importFn);
    return pkg.default || pkg;
  }
  return StitchingMerger;
}
