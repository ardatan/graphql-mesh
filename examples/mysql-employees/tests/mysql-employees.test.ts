import { basename, join } from 'path';
import { findAndParseConfig } from '@graphql-mesh/cli';
import { ProcessedConfig } from '@graphql-mesh/config';
import { getMesh, MeshInstance } from '@graphql-mesh/runtime';
import { printSchemaWithDirectives } from '@graphql-tools/utils';

jest.setTimeout(30000);

describe('MySQL Employees', () => {
  let mesh: MeshInstance;
  let config: ProcessedConfig;
  beforeAll(async () => {
    config = await findAndParseConfig({
      dir: join(__dirname, '..'),
    });
    mesh = await getMesh(config);
  });
  it('should generate correct schema', async () => {
    expect(printSchemaWithDirectives(mesh.schema)).toMatchSnapshot('schema');
  });
  it('should give correct response for example queries', () =>
    Promise.all(
      config.documents.map(async operation => {
        const result = await mesh.execute(operation.document!);
        expect(result).toMatchSnapshot(
          basename(operation.location || 'anonymous') + '-query-result',
        );
      }),
    ));
  afterAll(() => {
    mesh?.destroy();
  });
});
