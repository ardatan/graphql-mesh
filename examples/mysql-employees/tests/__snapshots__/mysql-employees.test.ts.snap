// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MySQL Employees should generate correct schema: schema 1`] = `
"schema @transport(subgraph: "Employees", kind: "mysql", location: "mysql://root:passwd@localhost:33306/employees") {
  query: Query
  mutation: Mutation
}

directive @transport(subgraph: String, kind: String, location: String) repeatable on SCHEMA

directive @mysqlSelect(subgraph: String, table: String, columnMap: [[String]]) on FIELD_DEFINITION

directive @mysqlInsert(subgraph: String, table: String, primaryKeys: [String]) on FIELD_DEFINITION

directive @mysqlUpdate(subgraph: String, table: String, columnMap: [[String]]) on FIELD_DEFINITION

directive @mysqlDelete(subgraph: String, table: String) on FIELD_DEFINITION

directive @mysqlTableForeign(subgraph: String, columnName: String) on FIELD_DEFINITION

directive @mysqlCount(subgraph: String, table: String) on FIELD_DEFINITION

type Query {
  current_dept_emp(limit: Int, offset: Int, where: current_dept_emp_WhereInput, orderBy: current_dept_emp_OrderByInput): [current_dept_emp] @mysqlSelect(subgraph: "Employees", table: "current_dept_emp")
  count_current_dept_emp(where: current_dept_emp_WhereInput): Int @mysqlCount(subgraph: "Employees", table: "current_dept_emp")
  departments(limit: Int, offset: Int, where: departments_WhereInput, orderBy: departments_OrderByInput): [departments] @mysqlSelect(subgraph: "Employees", table: "departments")
  count_departments(where: departments_WhereInput): Int @mysqlCount(subgraph: "Employees", table: "departments")
  dept_emp(limit: Int, offset: Int, where: dept_emp_WhereInput, orderBy: dept_emp_OrderByInput): [dept_emp] @mysqlSelect(subgraph: "Employees", table: "dept_emp")
  count_dept_emp(where: dept_emp_WhereInput): Int @mysqlCount(subgraph: "Employees", table: "dept_emp")
  dept_emp_latest_date(limit: Int, offset: Int, where: dept_emp_latest_date_WhereInput, orderBy: dept_emp_latest_date_OrderByInput): [dept_emp_latest_date] @mysqlSelect(subgraph: "Employees", table: "dept_emp_latest_date")
  count_dept_emp_latest_date(where: dept_emp_latest_date_WhereInput): Int @mysqlCount(subgraph: "Employees", table: "dept_emp_latest_date")
  dept_manager(limit: Int, offset: Int, where: dept_manager_WhereInput, orderBy: dept_manager_OrderByInput): [dept_manager] @mysqlSelect(subgraph: "Employees", table: "dept_manager")
  count_dept_manager(where: dept_manager_WhereInput): Int @mysqlCount(subgraph: "Employees", table: "dept_manager")
  employees(limit: Int, offset: Int, where: employees_WhereInput, orderBy: employees_OrderByInput): [employees] @mysqlSelect(subgraph: "Employees", table: "employees")
  count_employees(where: employees_WhereInput): Int @mysqlCount(subgraph: "Employees", table: "employees")
  salaries(limit: Int, offset: Int, where: salaries_WhereInput, orderBy: salaries_OrderByInput): [salaries] @mysqlSelect(subgraph: "Employees", table: "salaries")
  count_salaries(where: salaries_WhereInput): Int @mysqlCount(subgraph: "Employees", table: "salaries")
  titles(limit: Int, offset: Int, where: titles_WhereInput, orderBy: titles_OrderByInput): [titles] @mysqlSelect(subgraph: "Employees", table: "titles")
  count_titles(where: titles_WhereInput): Int @mysqlCount(subgraph: "Employees", table: "titles")
  v_full_departments(limit: Int, offset: Int, where: v_full_departments_WhereInput, orderBy: v_full_departments_OrderByInput): [v_full_departments] @mysqlSelect(subgraph: "Employees", table: "v_full_departments")
  count_v_full_departments(where: v_full_departments_WhereInput): Int @mysqlCount(subgraph: "Employees", table: "v_full_departments")
  v_full_employees(limit: Int, offset: Int, where: v_full_employees_WhereInput, orderBy: v_full_employees_OrderByInput): [v_full_employees] @mysqlSelect(subgraph: "Employees", table: "v_full_employees")
  count_v_full_employees(where: v_full_employees_WhereInput): Int @mysqlCount(subgraph: "Employees", table: "v_full_employees")
}

"""VIEW"""
type current_dept_emp {
  emp_no: Int!
  dept_no: String!
  from_date: Date
  to_date: Date
}

"""
A date string, such as 2007-12-03, compliant with the \`full-date\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar Date

"""VIEW"""
input current_dept_emp_WhereInput {
  emp_no: String
  dept_no: String
  from_date: String
  to_date: String
}

"""VIEW"""
input current_dept_emp_OrderByInput {
  emp_no: OrderBy
  dept_no: OrderBy
  from_date: OrderBy
  to_date: OrderBy
}

enum OrderBy {
  asc
  desc
}

type departments {
  dept_no: String!
  dept_name: String!
  dept_emp(limit: Int, offset: Int, where: dept_emp_WhereInput, orderBy: dept_emp_OrderByInput): [dept_emp] @mysqlSelect(subgraph: "Employees", table: "dept_emp", columnMap: [["dept_no", "dept_no"]])
  dept_manager(limit: Int, offset: Int, where: dept_manager_WhereInput, orderBy: dept_manager_OrderByInput): [dept_manager] @mysqlSelect(subgraph: "Employees", table: "dept_manager", columnMap: [["dept_no", "dept_no"]])
}

type dept_emp {
  emp_no: Int!
  dept_no: String!
  from_date: Date!
  to_date: Date!
  departments(where: departments_WhereInput, orderBy: departments_OrderByInput, limit: Int, offset: Int): [departments] @mysqlSelect(subgraph: "Employees", table: "departments", columnMap: [["dept_no", "dept_no"]]) @mysqlTableForeign(subgraph: "Employees", columnName: "dept_no")
  employees(where: employees_WhereInput, orderBy: employees_OrderByInput, limit: Int, offset: Int): [employees] @mysqlSelect(subgraph: "Employees", table: "employees", columnMap: [["emp_no", "emp_no"]]) @mysqlTableForeign(subgraph: "Employees", columnName: "emp_no")
}

input departments_WhereInput {
  dept_no: String
  dept_name: String
}

input departments_OrderByInput {
  dept_no: OrderBy
  dept_name: OrderBy
}

type employees {
  emp_no: Int!
  birth_date: Date!
  first_name: String!
  last_name: String!
  gender: employees_gender!
  hire_date: Date!
  dept_emp(limit: Int, offset: Int, where: dept_emp_WhereInput, orderBy: dept_emp_OrderByInput): [dept_emp] @mysqlSelect(subgraph: "Employees", table: "dept_emp", columnMap: [["emp_no", "emp_no"]])
  dept_manager(limit: Int, offset: Int, where: dept_manager_WhereInput, orderBy: dept_manager_OrderByInput): [dept_manager] @mysqlSelect(subgraph: "Employees", table: "dept_manager", columnMap: [["emp_no", "emp_no"]])
  salaries(limit: Int, offset: Int, where: salaries_WhereInput, orderBy: salaries_OrderByInput): [salaries] @mysqlSelect(subgraph: "Employees", table: "salaries", columnMap: [["emp_no", "emp_no"]])
  titles(limit: Int, offset: Int, where: titles_WhereInput, orderBy: titles_OrderByInput): [titles] @mysqlSelect(subgraph: "Employees", table: "titles", columnMap: [["emp_no", "emp_no"]])
}

enum employees_gender {
  M
  F
}

input dept_emp_WhereInput {
  emp_no: String
  dept_no: String
  from_date: String
  to_date: String
}

input dept_emp_OrderByInput {
  emp_no: OrderBy
  dept_no: OrderBy
  from_date: OrderBy
  to_date: OrderBy
}

type dept_manager {
  emp_no: Int!
  dept_no: String!
  from_date: Date!
  to_date: Date!
  departments(where: departments_WhereInput, orderBy: departments_OrderByInput, limit: Int, offset: Int): [departments] @mysqlSelect(subgraph: "Employees", table: "departments", columnMap: [["dept_no", "dept_no"]]) @mysqlTableForeign(subgraph: "Employees", columnName: "dept_no")
  employees(where: employees_WhereInput, orderBy: employees_OrderByInput, limit: Int, offset: Int): [employees] @mysqlSelect(subgraph: "Employees", table: "employees", columnMap: [["emp_no", "emp_no"]]) @mysqlTableForeign(subgraph: "Employees", columnName: "emp_no")
}

input employees_WhereInput {
  emp_no: String
  birth_date: String
  first_name: String
  last_name: String
  gender: String
  hire_date: String
}

input employees_OrderByInput {
  emp_no: OrderBy
  birth_date: OrderBy
  first_name: OrderBy
  last_name: OrderBy
  gender: OrderBy
  hire_date: OrderBy
}

input dept_manager_WhereInput {
  emp_no: String
  dept_no: String
  from_date: String
  to_date: String
}

input dept_manager_OrderByInput {
  emp_no: OrderBy
  dept_no: OrderBy
  from_date: OrderBy
  to_date: OrderBy
}

type salaries {
  emp_no: Int!
  salary: Int!
  from_date: Date!
  to_date: Date!
  employees(where: employees_WhereInput, orderBy: employees_OrderByInput, limit: Int, offset: Int): [employees] @mysqlSelect(subgraph: "Employees", table: "employees", columnMap: [["emp_no", "emp_no"]]) @mysqlTableForeign(subgraph: "Employees", columnName: "emp_no")
}

input salaries_WhereInput {
  emp_no: String
  salary: String
  from_date: String
  to_date: String
}

input salaries_OrderByInput {
  emp_no: OrderBy
  salary: OrderBy
  from_date: OrderBy
  to_date: OrderBy
}

type titles {
  emp_no: Int!
  title: String!
  from_date: Date!
  to_date: Date
  employees(where: employees_WhereInput, orderBy: employees_OrderByInput, limit: Int, offset: Int): [employees] @mysqlSelect(subgraph: "Employees", table: "employees", columnMap: [["emp_no", "emp_no"]]) @mysqlTableForeign(subgraph: "Employees", columnName: "emp_no")
}

input titles_WhereInput {
  emp_no: String
  title: String
  from_date: String
  to_date: String
}

input titles_OrderByInput {
  emp_no: OrderBy
  title: OrderBy
  from_date: OrderBy
  to_date: OrderBy
}

"""VIEW"""
type dept_emp_latest_date {
  emp_no: Int!
  from_date: Date
  to_date: Date
}

"""VIEW"""
input dept_emp_latest_date_WhereInput {
  emp_no: String
  from_date: String
  to_date: String
}

"""VIEW"""
input dept_emp_latest_date_OrderByInput {
  emp_no: OrderBy
  from_date: OrderBy
  to_date: OrderBy
}

"""VIEW"""
type v_full_departments {
  dept_no: String!
  dept_name: String!
  manager: String
}

"""VIEW"""
input v_full_departments_WhereInput {
  dept_no: String
  dept_name: String
  manager: String
}

"""VIEW"""
input v_full_departments_OrderByInput {
  dept_no: OrderBy
  dept_name: OrderBy
  manager: OrderBy
}

"""VIEW"""
type v_full_employees {
  emp_no: Int!
  first_name: String!
  last_name: String!
  birth_date: Date!
  gender: v_full_employees_gender!
  hire_date: Date!
  department: String
}

enum v_full_employees_gender {
  M
  F
}

"""VIEW"""
input v_full_employees_WhereInput {
  emp_no: String
  first_name: String
  last_name: String
  birth_date: String
  gender: String
  hire_date: String
  department: String
}

"""VIEW"""
input v_full_employees_OrderByInput {
  emp_no: OrderBy
  first_name: OrderBy
  last_name: OrderBy
  birth_date: OrderBy
  gender: OrderBy
  hire_date: OrderBy
  department: OrderBy
}

type Mutation {
  insert_current_dept_emp(current_dept_emp: current_dept_emp_InsertInput!): current_dept_emp @mysqlInsert(subgraph: "Employees", table: "current_dept_emp", primaryKeys: [])
  update_current_dept_emp(current_dept_emp: current_dept_emp_UpdateInput!, where: current_dept_emp_WhereInput): current_dept_emp @mysqlUpdate(subgraph: "Employees", table: "current_dept_emp")
  delete_current_dept_emp(where: current_dept_emp_WhereInput): Boolean @mysqlDelete(subgraph: "Employees", table: "current_dept_emp")
  insert_departments(departments: departments_InsertInput!): departments @mysqlInsert(subgraph: "Employees", table: "departments", primaryKeys: ["dept_no"])
  update_departments(departments: departments_UpdateInput!, where: departments_WhereInput): departments @mysqlUpdate(subgraph: "Employees", table: "departments")
  delete_departments(where: departments_WhereInput): Boolean @mysqlDelete(subgraph: "Employees", table: "departments")
  insert_dept_emp(dept_emp: dept_emp_InsertInput!): dept_emp @mysqlInsert(subgraph: "Employees", table: "dept_emp", primaryKeys: ["emp_no", "dept_no"])
  update_dept_emp(dept_emp: dept_emp_UpdateInput!, where: dept_emp_WhereInput): dept_emp @mysqlUpdate(subgraph: "Employees", table: "dept_emp")
  delete_dept_emp(where: dept_emp_WhereInput): Boolean @mysqlDelete(subgraph: "Employees", table: "dept_emp")
  insert_dept_emp_latest_date(dept_emp_latest_date: dept_emp_latest_date_InsertInput!): dept_emp_latest_date @mysqlInsert(subgraph: "Employees", table: "dept_emp_latest_date", primaryKeys: [])
  update_dept_emp_latest_date(dept_emp_latest_date: dept_emp_latest_date_UpdateInput!, where: dept_emp_latest_date_WhereInput): dept_emp_latest_date @mysqlUpdate(subgraph: "Employees", table: "dept_emp_latest_date")
  delete_dept_emp_latest_date(where: dept_emp_latest_date_WhereInput): Boolean @mysqlDelete(subgraph: "Employees", table: "dept_emp_latest_date")
  insert_dept_manager(dept_manager: dept_manager_InsertInput!): dept_manager @mysqlInsert(subgraph: "Employees", table: "dept_manager", primaryKeys: ["emp_no", "dept_no"])
  update_dept_manager(dept_manager: dept_manager_UpdateInput!, where: dept_manager_WhereInput): dept_manager @mysqlUpdate(subgraph: "Employees", table: "dept_manager")
  delete_dept_manager(where: dept_manager_WhereInput): Boolean @mysqlDelete(subgraph: "Employees", table: "dept_manager")
  insert_employees(employees: employees_InsertInput!): employees @mysqlInsert(subgraph: "Employees", table: "employees", primaryKeys: ["emp_no"])
  update_employees(employees: employees_UpdateInput!, where: employees_WhereInput): employees @mysqlUpdate(subgraph: "Employees", table: "employees")
  delete_employees(where: employees_WhereInput): Boolean @mysqlDelete(subgraph: "Employees", table: "employees")
  insert_salaries(salaries: salaries_InsertInput!): salaries @mysqlInsert(subgraph: "Employees", table: "salaries", primaryKeys: ["emp_no", "from_date"])
  update_salaries(salaries: salaries_UpdateInput!, where: salaries_WhereInput): salaries @mysqlUpdate(subgraph: "Employees", table: "salaries")
  delete_salaries(where: salaries_WhereInput): Boolean @mysqlDelete(subgraph: "Employees", table: "salaries")
  insert_titles(titles: titles_InsertInput!): titles @mysqlInsert(subgraph: "Employees", table: "titles", primaryKeys: ["emp_no", "title", "from_date"])
  update_titles(titles: titles_UpdateInput!, where: titles_WhereInput): titles @mysqlUpdate(subgraph: "Employees", table: "titles")
  delete_titles(where: titles_WhereInput): Boolean @mysqlDelete(subgraph: "Employees", table: "titles")
  insert_v_full_departments(v_full_departments: v_full_departments_InsertInput!): v_full_departments @mysqlInsert(subgraph: "Employees", table: "v_full_departments", primaryKeys: [])
  update_v_full_departments(v_full_departments: v_full_departments_UpdateInput!, where: v_full_departments_WhereInput): v_full_departments @mysqlUpdate(subgraph: "Employees", table: "v_full_departments")
  delete_v_full_departments(where: v_full_departments_WhereInput): Boolean @mysqlDelete(subgraph: "Employees", table: "v_full_departments")
  insert_v_full_employees(v_full_employees: v_full_employees_InsertInput!): v_full_employees @mysqlInsert(subgraph: "Employees", table: "v_full_employees", primaryKeys: [])
  update_v_full_employees(v_full_employees: v_full_employees_UpdateInput!, where: v_full_employees_WhereInput): v_full_employees @mysqlUpdate(subgraph: "Employees", table: "v_full_employees")
  delete_v_full_employees(where: v_full_employees_WhereInput): Boolean @mysqlDelete(subgraph: "Employees", table: "v_full_employees")
}

"""VIEW"""
input current_dept_emp_InsertInput {
  emp_no: Int!
  dept_no: String!
  from_date: Date
  to_date: Date
}

"""VIEW"""
input current_dept_emp_UpdateInput {
  emp_no: Int
  dept_no: String
  from_date: Date
  to_date: Date
}

input departments_InsertInput {
  dept_no: String!
  dept_name: String!
}

input departments_UpdateInput {
  dept_no: String
  dept_name: String
}

input dept_emp_InsertInput {
  emp_no: Int!
  dept_no: String!
  from_date: Date!
  to_date: Date!
}

input dept_emp_UpdateInput {
  emp_no: Int
  dept_no: String
  from_date: Date
  to_date: Date
}

"""VIEW"""
input dept_emp_latest_date_InsertInput {
  emp_no: Int!
  from_date: Date
  to_date: Date
}

"""VIEW"""
input dept_emp_latest_date_UpdateInput {
  emp_no: Int
  from_date: Date
  to_date: Date
}

input dept_manager_InsertInput {
  emp_no: Int!
  dept_no: String!
  from_date: Date!
  to_date: Date!
}

input dept_manager_UpdateInput {
  emp_no: Int
  dept_no: String
  from_date: Date
  to_date: Date
}

input employees_InsertInput {
  emp_no: Int!
  birth_date: Date!
  first_name: String!
  last_name: String!
  gender: employees_gender!
  hire_date: Date!
}

input employees_UpdateInput {
  emp_no: Int
  birth_date: Date
  first_name: String
  last_name: String
  gender: employees_gender
  hire_date: Date
}

input salaries_InsertInput {
  emp_no: Int!
  salary: Int!
  from_date: Date!
  to_date: Date!
}

input salaries_UpdateInput {
  emp_no: Int
  salary: Int
  from_date: Date
  to_date: Date
}

input titles_InsertInput {
  emp_no: Int!
  title: String!
  from_date: Date!
  to_date: Date
}

input titles_UpdateInput {
  emp_no: Int
  title: String
  from_date: Date
  to_date: Date
}

"""VIEW"""
input v_full_departments_InsertInput {
  dept_no: String!
  dept_name: String!
  manager: String
}

"""VIEW"""
input v_full_departments_UpdateInput {
  dept_no: String
  dept_name: String
  manager: String
}

"""VIEW"""
input v_full_employees_InsertInput {
  emp_no: Int!
  first_name: String!
  last_name: String!
  birth_date: Date!
  gender: v_full_employees_gender!
  hire_date: Date!
  department: String
}

"""VIEW"""
input v_full_employees_UpdateInput {
  emp_no: Int
  first_name: String
  last_name: String
  birth_date: Date
  gender: v_full_employees_gender
  hire_date: Date
  department: String
}"
`;

exports[`MySQL Employees should give correct response for example queries: example-query.graphql-query-result 1`] = `
{
  "data": {
    "employees": [
      {
        "__typename": "employees",
        "birth_date": "1953-09-02",
        "dept_emp": [
          {
            "departments": [
              {
                "dept_name": "Development",
                "dept_no": "d005",
              },
            ],
            "dept_no": "d005",
            "emp_no": 10001,
            "from_date": "1986-06-26",
            "to_date": "9999-01-01",
          },
        ],
        "emp_no": 10001,
        "first_name": "Georgi",
        "gender": "M",
        "hire_date": "1986-06-26",
        "last_name": "Facello",
      },
      {
        "__typename": "employees",
        "birth_date": "1964-06-02",
        "dept_emp": [
          {
            "departments": [
              {
                "dept_name": "Sales",
                "dept_no": "d007",
              },
            ],
            "dept_no": "d007",
            "emp_no": 10002,
            "from_date": "1996-08-03",
            "to_date": "9999-01-01",
          },
        ],
        "emp_no": 10002,
        "first_name": "Bezalel",
        "gender": "F",
        "hire_date": "1985-11-21",
        "last_name": "Simmel",
      },
      {
        "__typename": "employees",
        "birth_date": "1959-12-03",
        "dept_emp": [
          {
            "departments": [
              {
                "dept_name": "Production",
                "dept_no": "d004",
              },
            ],
            "dept_no": "d004",
            "emp_no": 10003,
            "from_date": "1995-12-03",
            "to_date": "9999-01-01",
          },
        ],
        "emp_no": 10003,
        "first_name": "Parto",
        "gender": "M",
        "hire_date": "1986-08-28",
        "last_name": "Bamford",
      },
      {
        "__typename": "employees",
        "birth_date": "1954-05-01",
        "dept_emp": [
          {
            "departments": [
              {
                "dept_name": "Production",
                "dept_no": "d004",
              },
            ],
            "dept_no": "d004",
            "emp_no": 10004,
            "from_date": "1986-12-01",
            "to_date": "9999-01-01",
          },
        ],
        "emp_no": 10004,
        "first_name": "Chirstian",
        "gender": "M",
        "hire_date": "1986-12-01",
        "last_name": "Koblick",
      },
      {
        "__typename": "employees",
        "birth_date": "1955-01-21",
        "dept_emp": [
          {
            "departments": [
              {
                "dept_name": "Human Resources",
                "dept_no": "d003",
              },
            ],
            "dept_no": "d003",
            "emp_no": 10005,
            "from_date": "1989-09-12",
            "to_date": "9999-01-01",
          },
        ],
        "emp_no": 10005,
        "first_name": "Kyoichi",
        "gender": "M",
        "hire_date": "1989-09-12",
        "last_name": "Maliniak",
      },
    ],
  },
}
`;
